### ISSUE TITLE
Question : evaluation after every training epoch

### ISSUE THREAD

**Issue by Yongtae723** [huggingface/setfit#178]
# Thank you 
Hello!
I am Yongtae, a senior ML engineer in japan.
Thank you for publishing a genuinely excellent paper and code.
Few-shot learning and multilingual support are appreciated by engineers like me who work abroad!

# Question
I felt this model easily overfit to train data if the number of epochs is over 2 or train data contains similar data.
Therefore I would like to evaluate the model after every training epoch to find out the best epoch number.

But as shown [here](https://github.com/huggingface/setfit/blob/99c30746799a09e0267427b8a7b8650568222b48/src/setfit/trainer.py#L363), it seems difficult to evaluate the model at every epoch, because the body part is trained on full epoch at the beginning of the training. 

so I would like to change like below

```python
for epoch in num_epochs:
    self.model.model_body.fit(
                train_objectives=[(train_dataloader, train_loss)],
                epochs=1,
                steps_per_epoch=train_steps,
                optimizer_params={"lr": learning_rate},
                warmup_steps=warmup_steps,
                show_progress_bar=True,
                use_amp=self.use_amp,
            )

    if not is_differentiable_head or not self._freeze:
            # Train the final classifier
            self.model.fit(
                x_train,
                y_train,
                num_epochs=1,
                batch_size=batch_size,
                learning_rate=learning_rate,
                body_learning_rate=body_learning_rate,
                l2_weight=l2_weight,
                show_progress_bar=True,
            )

    somehow_evaluete()
```

Does it make sense to you?
Or if I fork and make that change, are there any problem?

I am looking forward to your reply

Best and thank you in advance!

**Comment by tomaarsen** [comment]
Hello!

My understanding of the paper (as another enthusiast like yourself) is that the model is trained in two separate phases. i.e.:

1. The SentenceTransformer body is *fully* finetuned.
2. Only then, the classifier head is *fully* trained.

The approach that you're describing in your code block is to repeatedly finetune one epoch on the SentenceTransformer body and then train one epoch on the classifier head, which is not equivalent to the paper.

The best that may be implementable is a (list of) callback function(s) that can be provided via some argument to `SetFitTrainer`, which may be called after every epoch of either of the two training phases, à la [transformer](https://huggingface.co/docs/transformers/v4.24.0/en/main_classes/trainer#transformers.Trainer.callbacks) or [PyTorch Lightning](https://pytorch-lightning.readthedocs.io/en/stable/common/early_stopping.html) callbacks. However, this is a bit awkward due to the two separate training phases.

**Comment by Yongtae723** [comment]
Hi Tom!
Thank you for your reply!

I understand your point. However, it is still very difficult to evaluate the performance of the model at each epoch when training SentenceTransformer (since the model_head is not trained).

However, I have a good idea to solve this problem and am now implementing the training of setfit with PyTorch Lightning.
I am almost done with the coding and will contact you after it is published, so please give us a brief comment if possible.

I also expect that wrap in PyTorch Lighining will facilitate logging, use of callbacks, and management of training devices.

**Comment by Yongtae723** [comment]
Hi Tom!
I am so sorry for the late reply.
We finally finished coding and publish (here)[https://github.com/jxpress/seifit-pytorch-lightning].

Regarding evaluation after every epoch, See [here](https://github.com/jxpress/seifit-pytorch-lightning/blob/main/documents/Implemented_strategy.md)

if you have any opinion, feel free to give me a comment.
Also if you feel our implementation is good enough, could you introduce our repository in your README? 

thank you.
I hope you doing great

Yongtae

**Comment by tomaarsen** [comment]
Very interesting work! Well done. It certainly solves some problems that we currently have with the main work (callbacks, gradient accumulation, etc.). I'm open to mentioning the work in a "Related work" section of sorts in the README, but I'll have to leave that decision to @lewtun.

- Tom Aarsen

**Comment by Yongtae723** [comment]
Hi Tom!

Can I make PR to add mentioning to introduce our work in README?

Of course, I understand The final decision will be made by @lewtun.
Thank you

**Comment by tomaarsen** [comment]
As far as I'm concerned: Feel free!
